File: api/filing/models/filing_model.py
"""
Filing Model
===========

Complete filing model that contains all tax return data organized by section.
"""

from .api_base_model import ApiBaseModel
from typing import List

# Chapter VIA Deductions - Popular
from domain.filing.deductions.chapter_via_deduction import ChapterVIADeductions
from domain.filing.deductions.popular.models.deduction_80c_model import Deduction80CModel
from domain.filing.deductions.popular.models.deduction_80ccc_model import Deduction80CCCModel
from domain.filing.deductions.popular.models.deduction_80ccd1_model import Deduction80CCD1Model
from domain.filing.deductions.popular.models.deduction_80ccd1b_model import Deduction80CCD1BModel
from domain.filing.deductions.popular.models.deduction_80ccd2_model import Deduction80CCD2Model

# Chapter VIA Deductions - Donations
from domain.filing.deductions.donation.models.deduction_80g_model import Deduction80GModel
from domain.filing.deductions.donation.models.deduction_80gga_model import Deduction80GGAModel
from domain.filing.deductions.donation.models.deduction_80ggc_model import Deduction80GGCModel

# Chapter VIA Deductions - Loans
from domain.filing.deductions.loan.models.deduction_80e_model import Deduction80EModel
from domain.filing.deductions.loan.models.deduction_80ee_model import Deduction80EEModel
from domain.filing.deductions.loan.models.deduction_80eea_model import Deduction80EEAModel
from domain.filing.deductions.loan.models.deduction_80eeb_model import Deduction80EEBModel

# Chapter VIA Deductions - Medical
from domain.filing.deductions.medical.models.deduction_80d_model import Deduction80DModel
from domain.filing.deductions.medical.models.deduction_80dd_model import Deduction80DDModel
from domain.filing.deductions.medical.models.deduction_80ddb_model import Deduction80DDBModel
from domain.filing.deductions.medical.models.deduction_80u_model import Deduction80UModel

# Chapter VIA Deductions - Other
from domain.filing.deductions.other.models.deduction_80cch_model import Deduction80CCHModel
from domain.filing.deductions.other.models.deduction_80gg_model import Deduction80GGModel
from domain.filing.deductions.other.models.deduction_80qqb_model import Deduction80QQBModel
from domain.filing.deductions.other.models.deduction_80rrb_model import Deduction80RRBModel
from domain.filing.deductions.other.models.deduction_80tta_model import Deduction80TTAModel
from domain.filing.deductions.other.models.deduction_80ttb_model import Deduction80TTBModel
from domain.filing.deductions.other.models.other_deduction_model import OtherDeductionModel
# Income Sections
from domain.filing.income.salary.models.salary_model import SalaryModel
from domain.filing.income.house_property.models.property_model import PropertyModel
from domain.filing.income.other.interest_income.models.interest_income_model import InterestIncomeModel
from domain.filing.income.other.equity_compensation_income.models.equity_compensation_income_model import EquityCompensationIncomeModel
from domain.filing.income.other.dividend.models.dividend_income_model import DividendIncomeModel
from domain.filing.income.other.foreign_income.models.foreign_income_model import ForeignIncomeModel
from domain.filing.income.other.capital_gains.capital_gains_securities.models.cg_securities_model import CapitalGainsSecuritiesModel
from domain.filing.income.other.capital_gains.capital_gains_real_estate.models.realestate_model import RealEstateModel
from domain.filing.income.other.capital_gains.capital_gains_foreign.models.foreign_assets_model import ForeignCapitalGains
from domain.filing.income.other.capital_gains.capital_gains_movable.models.movable_capital_gains_wrapper_model import MovableCapitalGainsWithImprovements
from domain.filing.income.other.capital_gains.capital_deemed.models.deemed_capital_gains_wrapper_model import DeemedCapitalGainsWrapper
from domain.filing.income.professional_income.models import ProfessionalIncomeModel
from domain.filing.income.other.agricultural_income.models.agricultural_income_model import AgriculturalIncomeModel
from domain.filing.income.other.crypto_vda_income.models.crypto_vda_income_model import CryptoVDAIncomeModel
from domain.filing.income.other.lottery_online_gaming_income.models.lottery_online_gaming_income_model import LotteryOnlineGamingIncomeModel

# Personal Details
from domain.filing.person.personal_details.models.person_model import PersonModel
from domain.filing.person.person_address.models.address_model import AddressModel
from domain.filing.person.bank_account_details.models.bank_account_model import BankAccountModel 

# Tax Credits
from domain.filing.tax_calculation.models.tax_calculation_response import TaxCalculationResponseModel
from domain.filing.tax_credit.models.tds_model import TDSModel
from domain.filing.tax_credit.models.tcs_model import TCSModel
from domain.filing.tax_credit.models.tax_paid_self_model import TaxPaidSelfModel

# Validation
from domain.filing.itr.validations.models.validation import ValidationError

# Form16 Metadata
from domain.filing.income.form16.models.form16_metadata import Form16Metadata

# Assets & Liabilities
from domain.filing.assets_and_liabilities.immovable_assets.models.immovable_assets_model import ImmovableAssetsModel
from domain.filing.assets_and_liabilities.financial_assets.models.financial_assets_model import FinancialAssetsModel
from domain.filing.assets_and_liabilities.other_assets.models.other_assets_model import OtherAssetsModel
from domain.filing.assets_and_liabilities.liabilities.models.liabilities_model import LiabilitiesModel
from domain.filing.assets_and_liabilities.investment_firm_llp_aop.models.investment_firm_llp_aop_model import InvestmentFirmLlpAopModel


class FilingModel(ApiBaseModel):
    """Complete filing model containing all tax return sections."""
    
    # Filing metadata
    filing_id: int
    assessment_year: str | None = None
    regime: str | None = None
    
    ############# Personal Details #############
    
    person: PersonModel | None = None
    person_address: AddressModel | None = None
    bank_account: List[BankAccountModel] = []

    ############# Deduction Sections #############
    
    # Popular Sections
    section_80c: List[Deduction80CModel] = []
    section_80ccc: List[Deduction80CCCModel] = []
    section_80ccd1: List[Deduction80CCD1Model] = []
    section_80ccd1b: List[Deduction80CCD1BModel] = []
    section_80ccd2: List[Deduction80CCD2Model] = []
    
    # Donations
    section_80g: List[Deduction80GModel] = []
    section_80gga: List[Deduction80GGAModel] = []
    section_80ggc: List[Deduction80GGCModel] = []
    
    # Loans
    section_80e: List[Deduction80EModel] = []
    section_80ee: Deduction80EEModel | None = None
    section_80eea: Deduction80EEAModel | None = None
    section_80eeb: Deduction80EEBModel | None = None
    
    # Medical
    section_80d: Deduction80DModel | None = None
    section_80dd: Deduction80DDModel | None = None
    section_80ddb: Deduction80DDBModel | None = None
    section_80u: Deduction80UModel | None = None
    
    # Other Deductions
    section_80cch: Deduction80CCHModel | None = None
    section_80gg: Deduction80GGModel | None = None
    section_80qqb: Deduction80QQBModel | None = None
    section_80rrb: Deduction80RRBModel | None = None
    section_80tta: Deduction80TTAModel | None = None
    section_80ttb: Deduction80TTBModel | None = None
    other_deductions: OtherDeductionModel | None = None
    ############# Income Sections #############
    
    salary: List[SalaryModel] = []
    house_property: List[PropertyModel] = []
    professional_income: ProfessionalIncomeModel | None = None
    equity_compensation_income: EquityCompensationIncomeModel | None = None

    # Other Income
    interest_income: List[InterestIncomeModel] = []
    dividend_income: DividendIncomeModel | None = None
    foreign_income: ForeignIncomeModel | None = None
    capital_gains_securities: CapitalGainsSecuritiesModel | None = None
    capital_gains_real_estate: RealEstateModel | None = None
    
    # Separated capital gains - each type independently managed
    capital_gains_foreign: List[ForeignCapitalGains] = []
    capital_gains_movable: MovableCapitalGainsWithImprovements | None = None
    capital_gains_deemed: DeemedCapitalGainsWrapper | None = None
    
    
    # Special income buckets (generally Schedule OS / special rates)
    crypto_vda_income: CryptoVDAIncomeModel | None = None
    lottery_online_gaming_income: LotteryOnlineGamingIncomeModel | None = None

    # Agricultural Income
    agricultural_income: AgriculturalIncomeModel | None = None
    
    ############# Tax Credits #############
    
    tds: List[TDSModel] = []
    tcs: List[TCSModel] = []
    advance_tax: List[TaxPaidSelfModel] = []  
    
    ############# Form16 Metadata #############
    
    form16_metadata: List[Form16Metadata] = []

    ############# tax Calculation #############
    tax_computation: TaxCalculationResponseModel | None = None

    ############# User Validation Errors #############
    ############# Assets & Liabilities #############

    immovable_assets: List[ImmovableAssetsModel] = []
    financial_assets: List[FinancialAssetsModel] = []
    other_assets: List[OtherAssetsModel] = []
    liabilities: List[LiabilitiesModel] = []
    investment_firm_llp_aop: List[InvestmentFirmLlpAopModel] = []
    # Total penal interest + late fee (234A/234B/234C/234F) computed during ITR build
    tax_intrest: int = 0
    # User validation errors - for left panel (incomplete/invalid user input)
    user_validation_errors: List[ValidationError] = []

    # -----------------------------
    # Computed views for ITR mappers
    # -----------------------------
    
    chapterVIADeductions: ChapterVIADeductions = ChapterVIADeductions()
    
    def has_income(self) -> bool:
        """Check if filing has any actual income data (salary, house property, etc.)."""
        if self.salary:
            return True
        if self.house_property:
            return True
        if self.professional_income is not None:
            return True
        if self.equity_compensation_income is not None:
            return True
        if self.interest_income:
            return True
        if self.dividend_income is not None:
            return True
        if self.foreign_income is not None:
            return True
        return False
    
# FilingModel

File: web/filing/models/filing-model.ts
/**
 * Filing Model - Root Level
 * Complete hierarchical structure of the FilingModel
 * All field names use camelCase convention
 */

// Personal Details
import { PersonDetailsModel } from './person/person-details-model'
import { AddressModel } from './person/person-address-model'
import { BankAccountModel } from './person/bank-account-model'

// Deductions - Popular
import { Deduction80CModel } from './deductions/popular/deduction-80c-model'
import { Deduction80CCCModel } from './deductions/popular/deduction-80ccc-model'
import { Deduction80CCD1Model } from './deductions/popular/deduction-80ccd1-model'
import { Deduction80CCD1BModel } from './deductions/popular/deduction-80ccd1b-model'
import { Deduction80CCD2Model } from './deductions/popular/deduction-80ccd2-model'

// Deductions - Donations
import { Deduction80GModel } from './deductions/donation/deduction-80g-model'
import { Deduction80GGAModel } from './deductions/donation/deduction-80gga-model'
import { Deduction80GGCModel } from './deductions/donation/deduction-80ggc-model'

// Deductions - Loans
import { Deduction80EModel } from './deductions/loan/deduction-80e-model'
import { Deduction80EEModel } from './deductions/loan/deduction-80ee-model'
import { Deduction80EEAModel } from './deductions/loan/deduction-80eea-model'
import { Deduction80EEBModel } from './deductions/loan/deduction-80eeb-model'

// Deductions - Medical
import { Deduction80DModel } from './deductions/medical/deduction-80d-model'
import { Deduction80DDModel } from './deductions/medical/deduction-80dd-model'
import { Deduction80DDBModel } from './deductions/medical/deduction-80ddb-model'
import { Deduction80UModel } from './deductions/medical/deduction-80u-model'

// Deductions - Other
import { Deduction80CCHModel } from './deductions/other/deduction-80cch-model'
import { Deduction80GGModel } from './deductions/other/deduction-80gg-model'
import { Deduction80QQBModel } from './deductions/other/deduction-80qqb-model'
import { Deduction80RRBModel } from './deductions/other/deduction-80rrb-model'
import { Deduction80TTAModel } from './deductions/other/deduction-80tta-model'
import { Deduction80TTBModel } from './deductions/other/deduction-80ttb-model'
import { OtherDeductionModel } from './deductions/other/other-deduction-model'

// Income
import { SalaryModel } from './income/salary/salary-model'
import { PropertyModel } from './income/house-property/property-model'
import { CapitalGainsSecuritiesModel } from './income/capital-gains/capital-gains-securities-model'
import { CapitalGainsRealEstateModel } from './income/capital-gains/cg-real-estate-property-model'
import { ForeignCapitalGains } from './income/capital-gains/foreign-capital-gains-model'
import { MovableCapitalGainsWithImprovements } from './income/capital-gains/movable-capital-gains-wrapper-model'
import { DeemedCapitalGainsModel } from './income/capital-gains/deemed-capital-gains-model'

// Tax Credits
import { TDSModel } from './tax-credits/tds-model'
import { TCSModel } from './tax-credits/tcs-model'
import { TaxPaidSelfModel } from './tax-credits/tax-paid-self-model'
import { TaxComputationModel } from './tax-computation-model'
import type { ImmovableAsset } from './assets/immovable-asset-model'
import type { OtherAsset } from './assets/other-asset-model'
import type { FinancialAsset } from './assets/financial-asset-model'
import type { Liability } from './assets/liability-model'
import type { PartnershipInvestment } from './assets/partnership-investment-model'
import { InterestIncomeModel } from './income/interest-income-model'
import { EquityCompensationIncomeModel } from './income/dividend/equity-compensation-income-model'
import { DividendIncomeModel } from './income/dividend/dividend-income-model'
import { ExemptIncomeModel } from './income/exempt-income-model'
import { LotteryOnlineGamingAggregatedModel as LotteryOnlineGamingModel } from './income/lottery-online-gaming-wrapper-model'
import { AgriculturalIncomeModel } from './income/agricultural-income-model'
import { ForeignIncomeModel } from './income/foreign-income-model'

// Form16 Metadata
export interface Form16Metadata {
  filingId: number
  form16UploadId?: number | null
  certificateNumber?: string | null
  processingDate?: string | null
  assessmentYear?: string | null
  citName?: string | null
  citAddressLine1?: string | null
  citAddressLine2?: string | null
  partATotalTaxDeposited: number
  verificationAuthorizedPerson?: string | null
  verificationFatherName?: string | null
  verificationDesignation?: string | null
  verificationPlace?: string | null
  verificationDate?: string | null
  verificationFullName?: string | null
}


/**
 * Validation Error Model - for backend validation errors
 */
export interface ValidationErrorModel {
  field: string
  message: string
  entityId?: number
}

/**
 * Complete Filing Model
 */
export interface FilingModel {
  // ############# Metadata #############
  filingId: number
  assessmentYear?: string | null
  regime?: string | null

  // ############# Personal Details #############
  person?: PersonDetailsModel | null
  personAddress?: AddressModel | null
  bankAccount: BankAccountModel[]

  // ############# Deduction Sections - Popular #############
  section80C: Deduction80CModel[]
  section80Ccc: Deduction80CCCModel[]
  section80Ccd1: Deduction80CCD1Model[]
  section80Ccd1B: Deduction80CCD1BModel[]
  section80Ccd2: Deduction80CCD2Model[]

  // ############# Deduction Sections - Donations #############
  section80G: Deduction80GModel[]
  section80Gga: Deduction80GGAModel[]
  section80Ggc: Deduction80GGCModel[]

  // ############# Deduction Sections - Loans #############
  section80E: Deduction80EModel[]
  section80Ee?: Deduction80EEModel | null
  section80Eea?: Deduction80EEAModel | null
  section80Eeb?: Deduction80EEBModel | null

  // ############# Deduction Sections - Medical #############
  section80D?: Deduction80DModel | null
  section80Dd?: Deduction80DDModel | null
  section80Ddb?: Deduction80DDBModel | null
  section80U?: Deduction80UModel | null

  // ############# Deduction Sections - Other #############
  section80Cch?: Deduction80CCHModel | null
  section80Gg?: Deduction80GGModel | null
  section80Qqb?: Deduction80QQBModel | null
  section80Rrb?: Deduction80RRBModel | null
  section80Tta?: Deduction80TTAModel | null
  section80Ttb?: Deduction80TTBModel | null
  otherDeductions?: OtherDeductionModel | null
  // ############# Income Sections #############
  salary: SalaryModel[]
  houseProperty: PropertyModel[]
  interestIncome: InterestIncomeModel[]
  equityCompensationIncome?: EquityCompensationIncomeModel | null
  foreignIncome?: ForeignIncomeModel | null
  capitalGainsSecurities?: CapitalGainsSecuritiesModel | null
  agriculturalIncome?: AgriculturalIncomeModel | null
  capitalGainsRealEstate?: CapitalGainsRealEstateModel | null
  capitalGainsForeign?: ForeignCapitalGains[] | null
  capitalGainsMovable?: MovableCapitalGainsWithImprovements | null
  capitalGainsDeemed?: DeemedCapitalGainsModel | null

  // ############# Tax Credits #############
  tds: TDSModel[]
  tcs: TCSModel[]
  advanceTax: TaxPaidSelfModel[]

  // ############# Dividend Income #############
  dividendIncome?: DividendIncomeModel | null
  exemptIncome?: ExemptIncomeModel | null
  lotteryOnlineGaming?: LotteryOnlineGamingModel | null

  // ############# Assets and Liabilities #############
  immovableAssets?: ImmovableAsset[] | null
  otherAssets?: OtherAsset[] | null
  financialAssets?: FinancialAsset[] | null
  liabilities?: Liability[] | null
  investmentFirmLlpAop?: PartnershipInvestment[] | null

  // ############# Form16 Metadata #############
  form16Metadata: Form16Metadata[]
  taxComputation?: TaxComputationModel | null

  // Total penal interest + late fee (234A/234B/234C/234F) from ITR build
  taxIntrest?: number

  // ############# User Validation Errors #############
  // User validation errors - for left panel (incomplete/invalid user input)
  userValidationErrors?: ValidationErrorModel[]

  // ############# Chapter VIA Deductions Breakdown #############
  chapterVIADeductionsNew?: ChapterVIADeductions | null
  chapterVIADeductionsOld?: ChapterVIADeductions | null
}

export interface DeductionAmount {
  claimed: number
  maxAllowed: number
  allowed: number
}

export interface ChapterVIADeductions {
  section80C?: DeductionAmount
  section80Ccc?: DeductionAmount
  section80Ccd1?: DeductionAmount
  section80Ccd1B?: DeductionAmount
  section80Ccd2?: DeductionAmount
  section80D?: DeductionAmount
  section80Dd?: DeductionAmount
  section80Ddb?: DeductionAmount
  section80E?: DeductionAmount
  section80Ee?: DeductionAmount
  section80Eea?: DeductionAmount
  section80Eeb?: DeductionAmount
  section80Tta?: DeductionAmount
  section80Ttb?: DeductionAmount
  section80G?: DeductionAmount
  section80Gg?: DeductionAmount
  section80Gga?: DeductionAmount
  section80Ggc?: DeductionAmount
  section80U?: DeductionAmount
  section80Cch?: DeductionAmount
  section80Qqb?: DeductionAmount
  section80Rrb?: DeductionAmount
}
